### PR TITLE
[MRG] Refactor creation of network 'feeds' (to allow new mechanism for adding network drives)

### LINKED ISSUE #104: clean up handling of external 'input' feeds
While working on #103 (after #98 broke the alpha example) we got hit by inconsistent handling of 'common' and 'unique' (renamed in #103) parameter sets (see discussion there for context).

The (new) `test_feed.py:test_extfeed()` illustrates the confusion: Poisson and Gaussian input parameters are always created by `create_pext`, but the 'common' ones aren't.

Variable naming (`git grep p_ext`) is quite poor, but rather than fiddling with names, it seems time to refactor the parameter handling process.

I suppose we're talking about an internal refactoring, where the API is less important than code clarity? Users shouldn't need to know anything about how the parameters are handled internally, right? The statefulness of the `ExtFeed` object is currently not used, so is it time to re-write the methods as functions? Or is there a higher meaning that's not obvious?

Some points to discuss here:
- [ ] should parameter-files be flat, or more structured? (backwards compatibility-issue vs. clarity)
- [ ]  per-need inclusion of external feeds into `net`, which the `gid_dict` would then reflect (currently inconsistent for common vs. unique types, where no common input gids are created in the network object when `t0 > tstop`)

### PR DESCRIPTION
**Pull request by cjayb**
Based on the work done in #177, continue creating new API for adding (and activating/simulating) external drives of the network (currently 'feeds'). Some ideas are discussed in a separate Doc (contact @cjayb for access). This is going to be a test-breaking PR due to very idiosyncratic handling of random seeds used in HNN-GUI feed assignment.

Edit 6 Nov:
Discussed GID-handling with @jasmainak and figured much could be won by disentangling it from NetworkBuilder. As it stands, `Network` creates the `gid_dict` on initialisation, so all that's really left for `Builder` to do is assign them to the different nodes (ranks).

Aims
- [x] clarify GID creation & handling
- [x] move all `ExtFeed` calls into `Network`, allowing instantiation of all external drives before `NetworkBuilder` is called (check old "API" still works)
- [x] move `parallel_backends.py:_clone_and_simulate:L34-35` to `NetworkBuilder` in order to "provide jitter between trials" (`prng_seedcore` needs to be different)
  * turned out not to be needed
- [x] generally refactor `Network` and `NetworkBuilder` so that building a new mechanism easier (separate PR)

For future PR
- demonstrate API ideas to allow incrementally adding drives
- create new tests of `Network` event times
- create new visualisations of `Network` event times 
- discuss and settle on seed handling 
- probably more...

Closes #104

### CODE REVIEW

**Comment by cjayb** [comment]
Just getting off to a start to make @jasmainak happy ;) Your last MAINT to #177 really helped clear it up for me, thanks! Hope to push some API ideas in soon.

File: hnn_core/network.py
Comment (rythorpe) [inline hnn_core/network.py:264]: This is a great idea @cjayb and will allow the user to more flexibly create or destroy cells or feeds! Since `Network.gid_dict` and `Network.pos_dict` are so intimately tied, would it make sense to combine their assignment under a unified method?

File: hnn_core/network_builder.py
Comment (rythorpe) [inline hnn_core/network_builder.py:299]: In support of this change, I can't think of any reason why `n_times` needs to be an attribute of `Network`. @blakecaldwell are there any GUI-related considerations we're missing here?

Comment (blakecaldwell) [reply hnn_core/network_builder.py:299]: I agree, it doesn't seem necessary. Thanks for notifying me @rythorpe.

**Comment by jasmainak** [comment]
@cjayb do you want to pair program sometime for this?

and by pair program, I mean really code and not discuss ;-)

**Comment by cjayb** [comment]
Hey man, yes I think that would be a good idea! I'm 6 hours ahead, but I'm
free most evenings next week. How about Tuesday 3rd after 11 am/5 pm? I'm
assuming you can't vote ;) Weds and Thur also work, both from mid-afternoon
onwards. /C

**Comment by cjayb** [comment]
@jasmainak as a discussion-starter for tomorrow. Sorry about the git-hell, will prob. have to ditch this and start from scratch anyway! Bottom line: a lot harder to reach the type of API we were discussing. Some ideas here (like a curry'ed "generating function"), but looking forward to your insight :)

**Comment by cjayb** [comment]
As the commit says: WIP So no need to review just yet ;) There's a lot of cruft I need to weed out, but basically very confident that `ExtFeed` is now out of `NetworkBuilder` for good. I'll give a shout when I think it might make sense to take a closer peek.

**Comment by cjayb** [comment]
Awh crap, MPI is not happy about my changes. Will try to backtrack order changes in `NetworkBuilder`.

File: hnn_core/tests/test_parallel_backends.py
Comment (cjayb) [inline hnn_core/tests/test_parallel_backends.py:118]: I got this comment wrong (will fix in next commit): the test is to make sure that two trials have _different_ model prediction values! Have updated PR description to indicate where the problem lies. No easy solution in sight. Good news is that MPI tests now pass (single-trial runs produce the expected spike times).

Comment (cjayb) [reply hnn_core/tests/test_parallel_backends.py:118]: Well I found a way around it, with smaller edits than I had feared

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:175]: I thought "node" meant the different machines that you distribute your jobs to.

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:191]: document it in docstring not in comments :) it's more easy to find there.

File: hnn_core/network_builder.py
Comment (jasmainak) [inline hnn_core/network_builder.py:425]: how about:

```suggestion
    def _gid_to_pc(self):
```

descriptive enough? I don't want the name to be too vague ...

also maybe add an argument called `threshold` so we don't accidentally add more dependencies on `params` inside the function.

File: hnn_core/network_builder.py
Comment (jasmainak) [inline hnn_core/network_builder.py:272]: self.trial_idx ?

to avoid learning a new variable name

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:292]: ic -> idx

it's what we use as short for "index" elsewhere

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:270]: `instantiate_feeds` ?

I think names of python objects/variables are more precise than descriptive names. Because now I was wondering what is "feed events"

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:63]: is there a realistic usecase for being able to update cell gids? Otherwise, I would just hide the functionality. The less you expose, the better

**Comment by jasmainak** [comment]
@cjayb most of my comments are nitpicky about variable names.

I think the next step is to write tests for the new methods. This way you'll know --> a) they work as you expect b) they are modular enough to be part of tests

File: hnn_core/network_builder.py
Comment (jasmainak) [inline hnn_core/network_builder.py:308]: I think `_gid_assign` should happen after `_create_cells_and_feeds`, no?

File: hnn_core/network_builder.py
Comment (jasmainak) [inline hnn_core/network_builder.py:308]: I have been thinking a bit about what is inside `_gid_assign`. I think ideally `NetworkBuilder` should take `rank` in the `__init__` that is passed from `mpi_child.py` and then `_clone_and_simulate`. Then inside the `__init__` you do:

```py
self.rank = rank
```

and use `self.rank` inside `_gid_assign` instead of `_get_rank`. This will make it clear that `NetworkBuilder` is actually on a particular node. That way you will also limit nested calls to Neuron which we know can behave somewhat arbitrarily. Probably for another PR though

File: hnn_core/network.py
Comment (cjayb) [reply hnn_core/network.py:175]: Well I guess there are plenty of inconsistencies in the variable naming, as there are in the comments. Not sure how to address that, other than keep on cleaning up? What would you call the cells of the network? "Cells", I suppose, though in terms of a network, they are just the "nodes" of a graph.

File: hnn_core/network.py
Comment (cjayb) [reply hnn_core/network.py:191]: Totally. Here I'm using a really _bad_ variable name (`_tmp...`) to remind me to get rid of it! The whole idea of this series of PRs is to get rid of 'feeds' and move to 'drives'. Semantics aside, the variable here turned out to be a list-of-dict-of-list-of-list---not sure that's the way to go :) Much of the work here is in order to get the tests to pass, _i.e._, match 'HNN Classic' models.

File: hnn_core/cell.py
Comment (cjayb) [reply hnn_core/cell.py:63]: Well this is a little stronger than hiding, no? I'm not sure the `Cell` gid ever needs to be updated, certainly not by users! Once it gets set, it should raise all hell if anything tries to change it. However, it would be nice to be able to _get_ the value using `cell.gid`, hence the `@property`. I know this complicates the `_Cell` class a little, but may make reading downstream code clearer (see `network_builder:L431-440`)?

File: hnn_core/network_builder.py
Comment (cjayb) [reply hnn_core/network_builder.py:308]: To my surprise too: No! Turns out L398 is the key. The for-loop there must be on `self.net._gid_list`, which contains precisely those cells (gids) that exist on _this rank/host_. So the `_gid_assign` (note the round-robin for-loop) essentially 'picks out' the gids that belong to the present rank, after which `create_...` instantiates them _here_.

File: hnn_core/network_builder.py
Comment (cjayb) [reply hnn_core/network_builder.py:308]: I like it, will keep in mind, as I'm pretty sure it'll belong to this PR.

File: hnn_core/network_builder.py
Comment (cjayb) [reply hnn_core/network_builder.py:425]: +1 on name for now, might change once it becomes clearer how to implement drives later.
added the param, as it's trivially easy here (much harder in `_create_cells...`)

File: hnn_core/network.py
Comment (cjayb) [reply hnn_core/network.py:270]: OK. I'm also still wrapping my head around this. Was stoked that I managed to instantiate all `event_times`  prior to `NetworkBuilder`! Wasn't easy, but now `network_builder` no longer does _any_ random number generation at all.

**Comment by cjayb** [comment]
> I think the next step is to write tests for the new methods. This way you'll know --> a) they work as you expect b) they are modular enough to be part of tests

Thanks for the comments so far. I'm very much in favour of writing tests, but wonder if it might make sense to discuss where to go from here first?

This PR is already substantial, and it's a major step forward: no more `event_times` created in `NetworkBuilder`. NB now takes care of GID assignment, cell creation & connection etc. All 'feeds' are created in `Network`, so they can be e.g. visualised prior to building & solving the network itself.

How about I take a break here & add some tests to `test_network` on expected `event_times` when using the default/testing parameters? But kinda leave it at that? The reason is that I have no intention of leaving `network.py` in the sorry state it is in! I want to implement a totally new API for getting the `event_times` instantiated, so making tests for the current bloody mess seems like an effort misplaced?

My proposal is: change the scope of this PR to the present (plus a few more tests). Then merge & move the remaining aims to a new branch. WDYT?

**Comment by cjayb** [comment]
Um so the rebase was a mess:
https://github.com/cjayb/hnn-core/pull/new/network_drive_pr_rebase
OK to force-push here? I guess the review history for this PR will be lost?

Obviously, I think my changes to `cell` and `pyramidal` are improvements, such as being explicit about `override_params` in `__init__` etc. Note that with the last commit (https://github.com/jonescompneurolab/hnn-core/commit/34dbddbba7909c790fe3f0134f838e991e7f4dea), `_create_cells_and_feeds` and `_gids_to_parallel_context` no longer rely on `params`!

**Comment by cjayb** [comment]
Um so the rebase was a mess:
https://github.com/cjayb/hnn-core/pull/new/network_drive_pr_rebase
OK to force-push here? I guess the review history for this PR will be lost?

Obviously, I think my changes to `cell` and `pyramidal` are improvements, such as being explicit about `override_params` in `__init__` etc. Note that with the last commit (), `_create_cells_and_feeds` and `_gids_to_parallel_context` no longer rely on `params`!

**Comment by jasmainak** [comment]
Feel free to do a force push! I will take a look at the PR later in the
afternoon

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:154]: `L2_basket` you mean?

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:172]: let's just do it in this PR. Seems straightforward. `gid_ranges` ?

File: hnn_core/network_builder.py
Comment (jasmainak) [inline hnn_core/network_builder.py:415]: XXX is reserved for more specific comments. Usually, we go over them before a release and see if some of them can be knocked off easily.

**Comment by jasmainak** [comment]
@cjayb okay can we bring the PR to a mergeable state? I'd be in favor of moving forward in small steps. Some tests can wait for later

File: hnn_core/network.py
Comment (jasmainak) [reply hnn_core/network.py:178]: I understand, but I'd still avoid bad variable names and call it `_feed_event_times` :) Who knows maybe you get busy after this PR. Also I will insist on the docstring ;-)

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:257]: docstring missing for public method

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:260]: this is code smell. Was it already there?

File: hnn_core/network.py
Comment (jasmainak) [reply hnn_core/network.py:305]: Just to clarify, this will only add or remove keys, not modify the values, right? E.g., if I have `L2_Pyramidal: range(0, 100)` and `L5_Pyramidal: range(100, 200)`, do you imagine each of the key-value pairs to be updated when new cells are added?

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:299]: pass arguments explicitly?

**Comment by jasmainak** [comment]
@cjayb can you rebase and bring the PR to MRG state? I think we're going in the right direction !

File: hnn_core/network.py
Comment (cjayb) [reply hnn_core/network.py:154]: I do :)

File: hnn_core/network_builder.py
Comment (cjayb) [reply hnn_core/network_builder.py:415]: OK, I'll go through the ones I've left and decide either to do something about or remove.

File: hnn_core/network.py
Comment (cjayb) [reply hnn_core/network.py:172]: done

File: hnn_core/network.py
Comment (cjayb) [reply hnn_core/network.py:260]: Nothing was already here: all this was hidden deep in `network_builder` nested loops! I've added a docstring to mention the parameter-override `n_jobs` in `dipole.py:simulate` that makes this necessary.

File: hnn_core/network.py
Comment (cjayb) [reply hnn_core/network.py:257]: made it private, but added docstring anyway ;)

File: hnn_core/network.py
Comment (cjayb) [reply hnn_core/network.py:299]: -1 The reliance on `pos_dict` for all inference is weird and should be fixed. I think it's more than just changing the calling pattern of this method.

**Comment by jasmainak** [comment]
@cjayb looks like you have rebase conflicts. Could you fix them? I'm fetching your PR and looking at it in the meanwhile.

File: hnn_core/dipole.py
Comment (jasmainak) [inline hnn_core/dipole.py:45]: ughgh ... how will this work when you instantiate the feeds outside of `simulate_dipole` ? do you have any ideas?

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:253]: nitpick, but you should use """ instead of ''' for multiline docstrings

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:271]: we should rethink this approach now that a lot of `feed.py` code is simplified in #204 and we begin to understand what that code is doing. I tagged you in #113

File: hnn_core/network_builder.py
Comment (jasmainak) [reply hnn_core/network_builder.py:320]: But if cells are created _after_ the GIDs are assigned to different nodes, how will we allow users to add feeds? User code should not depend on how the GID assignment is done ...

File: hnn_core/network_builder.py
Comment (jasmainak) [inline hnn_core/network_builder.py:410]: Not sure ... separate PR :)

**Comment by jasmainak** [comment]
PR looks fine to me, happy to merge once CIs come green. @rythorpe do you want to take a look?

File: hnn_core/dipole.py
Comment (cjayb) [reply hnn_core/dipole.py:45]: Just checked: `n_trials` is only an argument to `simulate_dipole`. All other parts of code check `net.params['N_trials']` to figure out the same. So there are no other places to worry about wrt. instantiation of `event_times`.
I know this isn't pretty. I've made `n_trials` an explicit argument of `_instantiate_feeds` to make it clearer there's a dependency there.

File: hnn_core/network.py
Comment (cjayb) [reply hnn_core/network.py:271]: Why not merge #204 first, then rebase and fix this (more invasive) PR on that? I also predict  #113 will then just melt away :)

File: hnn_core/network_builder.py
Comment (cjayb) [reply hnn_core/network_builder.py:320]: They can't. But that should be OK, no? We can create pretty visualisations of the input to a `net` before it ever gets built. So seeing the effects of adding and removing feeds (histograms etc.) can be done interactively, `_gid_assign` is not run until `Builder`. Note that the reliance on `self.net._p_unique` is something that the future implementation could make prettier, but right now it works as it is.

File: hnn_core/network_builder.py
Comment (cjayb) [reply hnn_core/network_builder.py:320]: Wasn't hard, just added rank and n_hosts as attributes to `NetworkBuilder` objects.

Comment (cjayb) [reply hnn_core/network_builder.py:320]: I take it back: something exploded so I rolled it back. Separate PR!

**Comment by cjayb** [comment]
Had to squash to make rebasing feasible: #190 really made it hard!
The last commit (@da134c7) is quite invasive, so maybe @ntolley would like to take a look? Basically I don't see why `times` should be an attribute of a `Network`? I've moved it into `Spikes`, which seems to me to make more sense (though could be wrong?).
I feel quite strongly though about `times` being the same for all `n_trials`! I've made an in-line comment about it in `parallel_backends.py`, please take a look.

@jasmainak I still think #204 could be merged before this, but I don't feel like another rebase on this one, so hope we could move fast?

File: hnn_core/parallel_backends.py
Comment (cjayb) [inline hnn_core/parallel_backends.py:62]: Is there any reason different trials should be allowed to have different `times`?!

**Comment by jasmainak** [comment]
yes, let's move fast. But another set of eyes other than me would be good. @ntolley or @rythorpe perhaps?

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:451]: raise a proper `TypeError` ...

**Review by jasmainak — APPROVED** [review]

File: hnn_core/cell.py
Comment (rythorpe) [reply hnn_core/cell.py:64]: Hmm, I'm not sure why we need 2 gid attributes here (i.e., `_assigned_gid` and `gid`). Why can't we just have one attribute named `_ArtificialCell._gid` that is set to the arg `gid` using your decorated setter function, and is retrieved by your decorated property function? Unless I'm missing something nuanced in the logic, such a change would simplify the gid code in this class as well as subsequent cell classes.

Comment (cjayb) [reply hnn_core/cell.py:63]: ah, maybe a little convoluted there yes. I just pushing something now, which I instantly regretted :( Ignore first push, my proposal will be the second one! To answer your question: I'm proposing to protect `.gid` because we don't want to risk anything messing with a `Cell`'a gid once building has begun. Perhaps there's a better way of achieving this?

Comment (rythorpe) [reply hnn_core/cell.py:65]: What I'm imagining is something like

```python
def __init__(self, event_times, threshold, gid=None):
   ...
   self.gid = gid

@gid.setter
def gid(self, gid)
   if self.gid is None:
      self.gid = gid
   else:
      raise RuntimeError('Global ID for this cell already assigned!')
```

Comment (cjayb) [reply hnn_core/cell.py:65]: well, I'm not sure it's getting any better (7d16670)? If not, could you explain your thoughts further? I like the idea of a `Cell` sporting its `gid` for downstream use. There's some duplication in the code now, though, so if you agree on something like this,  might have to create a superclass for `_Cell` and `_ArtificialCell` to inherit from?

Comment (rythorpe) [reply hnn_core/cell.py:65]: Oh sorry, I didn't refresh my browser and made my previous comment before seeing your most recent commit. I'm just trying to eliminate the need for having two attributes that store the gid in order to make the logic more readable. Correct me if I'm wrong, but the only consideration here is that the gid of a given cell should only be assigned to a value other than `None` once. Since this logic can be contained purely within a gid setter function, I don't understand why we need a `self.gid` and `self._gid` attribute.

Comment (cjayb) [reply hnn_core/cell.py:65]: Same here with the browser refresh ;) I'm trying your proposal, but `test_cell` hangs? I would be surprised if you could do `self.gid = gid`, isn't that trying to call the 'getter' and setter at the same time?! I've seen the `_`-prefix in cases like this elsewhere, thought it as a `pattern`...

Comment (cjayb) [reply hnn_core/cell.py:65]: > Correct me if I'm wrong, but the only consideration here is that the gid of a given cell should only be assigned to a value other than `None` once.

That's my goal, yes.

File: hnn_core/network_builder.py
Comment (rythorpe) [reply hnn_core/network_builder.py:318]: Just to clarify, @cjayb it sounds like what you're saying is that all aspects of the feed/network that we intend the user to interact with should be defined prior to `NetworkBuilder`? If so, this makes sense since the nrn interface should be invisible to the user.

Comment (cjayb) [reply hnn_core/network_builder.py:318]: Yes. That wasn't really my goal but it mostly just became so when extracting the feed instantiation into `Network`. Now I think it's a good goal though, as you say ;)

Comment (cjayb) [reply hnn_core/network_builder.py:318]: Note that I'm not implying that GID handling is "solved" quite yet. Baby steps... One of them being the protected cell attribute.

File: hnn_core/cell.py
Comment (jasmainak) [reply hnn_core/cell.py:65]: let me shed some light here. You can use private attributes for stuff that you don't want the user to mess with. For example if the object is complex and updating one attribute actually affects the rest of the object. For instance if the object had an attribute called `sfreq` (sampling frequency) and another attribute called `times`, you would have to update both simultaneously. This is why it sometimes makes sense to protect certain attributes and update them only through special functions.

Comment (rythorpe) [reply hnn_core/cell.py:65]: > Same here with the browser refresh ;) I'm trying your proposal, but `test_cell` hangs? I would be surprised if you could do `self.gid = gid`, isn't that trying to call the 'getter' and setter at the same time?!

Shoot, you're right. I think your implementation in 7d16670 is our best bet.

File: hnn_core/parallel_backends.py
Comment (rythorpe) [reply hnn_core/parallel_backends.py:62]: I'm guessing that the goal here was to maintain consistent formatting between the primary attributes of the `Spikes` class, but you're right that this is probably unnecessary. Any considerations we're missing @ntolley?

**Review by rythorpe — APPROVED** [review]
Okay, I think I now have a decent grasp of this PR. (Sorry for being so absent; as soon as I get this manuscript I've been working on submitted I'll be much more present.) Your method for creating jittered feed times all within `Network.instantiate_feeds()` such that they can be referenced on a trial-by-trial basis in `NetworkBuilder` and across parallel backends will be super useful as we move forward. Thanks @cjayb!

**Comment by jasmainak** [comment]
It's in, thanks @cjayb ! :-) Let's merge #204 next so you can make your next PR on top of it

File: hnn_core/cell.py
Comment (cjayb) [reply hnn_core/cell.py:65]: Can always be changed to something smoother later.

**Comment by cjayb** [comment]
Cool! Thanks both, good to get this out of the way :) @jasmainak what's the best way I can help rebase #204?

**Comment by jasmainak** [comment]
@cjayb I am working on my rebase and realized that `trial_event_times` should have `dict()` as the outermost level. Then inside you can probably have numpy arrays instead of "list of list" ... might be the cleanest.

**Comment by cjayb** [comment]
A dict with which keys though? I don't have a strong opinion there so feel free to suggest a change in #204!

**Comment by jasmainak** [comment]
It should be dict (keys are the src_type) of list (gids) of list (n_trials) of list (n_events) ? I think it will help unify things related to trial jitter etc.

File: hnn_core/parallel_backends.py
Comment (ntolley) [reply hnn_core/parallel_backends.py:62]: Consistency was the main motivation, the way it is written should not allow for different `times` across trials. `times` is actually created from directly from `params` when the network is instantiated, this function is really just copying the data.

Not really a big deal to change and probably cleaner this way. Most important thing is for the examples to demonstrate how to use the variable in plotting.

**Pull request merged**
